### PR TITLE
Fix base class property description

### LINKED ISSUE #152: Property summary of derived properties from a base class is not included in document
For contract class like below
```c#
    public class BaseClass
    {
        /// <summary>
        /// Base Property
        /// </summary>
        public string BaseProperty { get; }
    }
    public class ChildClass : BaseClass
    {
        /// <summary>
        /// Child Property
        /// </summary>
        public string ChildProperty { get; }
    }
```
Currently the schema generated for Child class is like below which is missing the description for Base Property.

    "ChildClass": {
      "type": "object",
      "properties": {
        "ChildProperty": {
          "type": "string",
          "description": "Child Property"
        },
        "BaseProperty": {
          "type": "string"
        }
      }
    }

That is happening b/c in the MemberSummaryToSchemaDescriptionFilter we fetch all the property xelement from XML, and then fetch the class name from it to see if schema for it already exists or not and then populate the property description. But in case of Base class the property xelement is like P:BaseClass.BaseProperty, so when we fetch class name its BaseClass which doesn't exists in Schema as property is listed under Child class and can't think of a way that we can map these base/child relationship in that filter.

So proposing, we pass XML Documents to SchemaReferenceRegistry and since we have type and property information in there we can easily fetch the summary information from xml for a property and update schema accordingly, code snippet like below

```c#
var innerSchema = FindOrAddReference(propertyInfo.PropertyType);
var propertyMembers = new List<XElement>();

foreach (var xmlDocument in _schemaGenerationSettings.XmlDocuments)
{
            propertyMembers.AddRange( xmlDocument.XPathSelectElements( "//doc/members/member" )
                            .Where(
                                m => m.Attribute( KnownXmlStrings.Name ) != null &&
                                     m.Attribute( KnownXmlStrings.Name ).Value == $"P:{propertyInfo.DeclaringType.FullName}.{propertyInfo.Name}" ) );
}

if (propertyMembers.FirstOrDefault() != null)
{
                        innerSchema.Description = propertyMembers.First().Element( KnownXmlStrings.Summary )?.Value
                            .RemoveBlankLines();
}
```

@PerthCharern  @scott-lin thoughts?

### PR DESCRIPTION
**Pull request by Shwetap05**
Fixes #152

### CODE REVIEW

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/DocumentFilters/MemberSummaryToSchemaDescriptionFilter.cs [hunk 1]
```diff
@@ -58,23 +59,18 @@ public class MemberSummaryToSchemaDescriptionFilter : IDocumentFilter
...
-
-                var splitPropertyName = fullPropertyName.Split('.');
-
-                // Take everything before the last period and remove the "P:" prefix.
-                var className =
-                    string.Join(".", splitPropertyName.Take(splitPropertyName.Length - 1))
-                        .Substring(startIndex: 2);
+            var typeNames = propertyMembers.Attributes(KnownXmlStrings.Name).Select(i => string
+                .Join(".", i.Value.Split('.').Take(i.Value.Split('.').Length - 1))
+                .Substring(2)).Distinct().ToList();
```
Comment (scott-lin) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/DocumentFilters/MemberSummaryToSchemaDescriptionFilter.cs]: This is very hard to read and understand. Lots of parens and nesting going on.

Can we do any of the following or a mix to make it better? Feels free to use your judgement.
1. Break it up by adding more intermediate local variable assignments, which would help describe what is going on better via the variable names, or (2) add comments to describe the logic, or (3) use newlines and indentation to make the parts more readable.

Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/DocumentFilters/MemberSummaryToSchemaDescriptionFilter.cs]: sure will do 2) and 3)

---
In reply to: [218900909](https://github.com/Microsoft/OpenAPI.NET.CSharpAnnotations/pull/154#discussion_r218900909) [](ancestors = 218900909)

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/Extensions/TypeExtensions.cs [hunk 2]
```diff
@@ -56,6 +56,27 @@ public static class TypeExtensions
                 [typeof(object)] = () => new OpenApiSchema {Type = "object"}
             };
 
+        /// <summary>
+        /// Gets the base types for the provided type
+        /// </summary>
+        /// <param name="type">The type to fetch base types for.</param>
+        /// <returns></returns>
```
Comment (scott-lin) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/Extensions/TypeExtensions.cs]: >>< [](start = 20, length = 2)

nit, document

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/Extensions/TypeExtensions.cs [hunk 3]
```diff
@@ -56,6 +56,27 @@ public static class TypeExtensions
...
+        public static List<Type> GetBaseTypes(this Type type)
```
Comment (scott-lin) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/Extensions/TypeExtensions.cs]: >List [](start = 22, length = 4)

nit, IList

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/TypeFetcher.cs [hunk 4]
```diff
@@ -116,6 +117,21 @@ private static int ExtractNumberOfGenerics(string typeName)
             return IsGenericType(typeName) ? int.Parse(typeName.Split('`')[1], CultureInfo.CurrentCulture) : 0;
         }
 
+        /// <summary>
+        /// Get base types for provided type.
+        /// </summary>
+        /// <param name="type">The type to fetch base types for.</param>
+        /// <returns>The list of base types.</returns>
+        public List<Type> GetBaseTypes(Type type)
```
Comment (scott-lin) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/TypeFetcher.cs]: >List [](start = 15, length = 4)

nit, IList

**Pull request merged**